lib/runner.js

```diff
@@ -984,66 +985,71 @@ Runner.prototype._uncaught = function(err) {
  * @param {{files: string[], options: Options}} [opts] - For subclasses
  * @return {Runner} Runner instance.
  */
-Runner.prototype.run = function(fn, opts) {
-  var self = this;
+Runner.prototype.run = function(fn, opts = {}) {
   var rootSuite = this.suite;
+  var options = opts.options || {};
 
+  debug('run(): got options: %O', options);
   fn = fn || function() {};
 
-  function start() {
+  const end = () => {
+    debug('run(): root suite completed; emitting %s', constants.EVENT_RUN_END);
+    this.emit(constants.EVENT_RUN_END);
+  };
+
+  const begin = () => {
+    debug('run(): emitting %s', constants.EVENT_RUN_BEGIN);
+    this.emit(constants.EVENT_RUN_BEGIN);
+    debug('run(): emitted %s', constants.EVENT_RUN_BEGIN);
+
+    this.runSuite(rootSuite, async () => {
+      end();
+    });
+  };
+
+  const prepare = () => {
     debug('run(): starting');
     // If there is an `only` filter
     if (rootSuite.hasOnly()) {
       rootSuite.filterOnly();
       debug('run(): filtered exclusive Runnables');
     }
-    self.state = constants.STATE_RUNNING;
-    if (self._delay) {
-      self.emit(constants.EVENT_DELAY_END);
+    this.state = constants.STATE_RUNNING;
+    if (this._delay) {
+      this.emit(constants.EVENT_DELAY_END);
       debug('run(): "delay" ended');
     }
-    debug('run(): emitting %s', constants.EVENT_RUN_BEGIN);
-    self.emit(constants.EVENT_RUN_BEGIN);
-    debug('run(): emitted %s', constants.EVENT_RUN_BEGIN);
 
-    self.runSuite(rootSuite, function() {
-      debug(
-        'run(): root suite completed; emitting %s',
-        constants.EVENT_RUN_END
-      );
-      self.emit(constants.EVENT_RUN_END);
-      debug('run(): emitted %s', constants.EVENT_RUN_END);
-    });
-  }
+    return begin();
+  };
 
   // references cleanup to avoid memory leaks
   if (this._opts.cleanReferencesAfterRun) {
-    this.on(constants.EVENT_SUITE_END, function(suite) {
+    this.on(constants.EVENT_SUITE_END, suite => {
       suite.cleanReferences();
     });
   }
 
   // callback
   this.on(constants.EVENT_RUN_END, function() {
-    self.state = constants.STATE_STOPPED;
-    debug(constants.EVENT_RUN_END);
+    this.state = constants.STATE_STOPPED;
     debug('run(): emitted %s', constants.EVENT_RUN_END);
-    fn(self.failures);
+    fn(this.failures);
   });
 
-  self._removeEventListener(process, 'uncaughtException', self.uncaught);
-  self._addEventListener(process, 'uncaughtException', self.uncaught);
+  this._removeEventListener(process, 'uncaughtException', this.uncaught);
+  this._removeEventListener(process, 'unhandledRejection', this.uncaught);
```

Comment:
this may have unintended consequences.  but we now have a global `unhandledRejection` listener.

I think it probably wants some integration tests.